# Ejercicios: Parte I    
Indicaciones: Los ejercicios 1, 2, 3 y 4 deben leer su data de links en Github y el código y los resultados deben ser publicados en páginas en Github.

## Ejercicio 1
1. Use two maps of points.

2. Compute the distance matrix for both maps.

3. Select one row of the distance matrix, and plot the two points with the minimal distance on top of the country of your choosing.

**1.1. Mapas de puntos a emplear:** 
- Plantas de energía solar en China (China3415_SolarPowerPlants)
- Plantas de energía eólica en China (China3415_WindPowerPlants)

- Ambos se encuentran proyectados y presentan un CRS de 3415.

Fuente empleada: [Global Power Plant Database](https://datasets.wri.org/dataset/globalpowerplantdatabase). Estos datos fueron preprocesados en otro Jupyter Notebook.


In [290]:
import geopandas as gpd
import os

China3415_PowerPlants = "https://github.com/Ciencia-de-datos-espaciales-2023-2/Evaluacion4_1ra-Parte_Mining_GeoDFs/raw/main/maps/China_3415/China3415_PowerPlant.gpkg"

China3415_SolarPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Solar")
China3415_WindPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Wind")

**1.2. Matriz de distancias entre ambos mapas**

Nota: Ambos mapas presentan proyecciones en metros. Pero, la matriz de distancias se encuentra en unidades de kilómetros y recopila distancias entre una planta de energía eólica a una de energía solar.

In [291]:
China3415_SolarPowerPlants_n = China3415_SolarPowerPlants.set_index("SolarPowerPlant_Name")
China3415_WindPowerPlants_n = China3415_WindPowerPlants.set_index("WindPowerPlant_Name")

DistanceMatrix_WindSolarPlants = China3415_WindPowerPlants_n.geometry.apply\
(lambda g: China3415_SolarPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 5 primeras filas de la matriz de distancias:
DistanceMatrix_WindSolarPlants.head()

SolarPowerPlant_Name                              Aksai A       Aksu 1  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             1990.422662  3181.257451   
Anhui Laian Baoshan Wind                      2354.174981  3669.157241   
Anhui Laian Longtougang Wind                  2366.436085  3681.410896   
Anhui Laian Longwosi Wind                     2355.954444  3670.937096   
Anning River Canyon (Phase 1) Dechang County  1572.310161  2679.343247   

SolarPowerPlant_Name                               Aksu A       Aksu B  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3182.821379  3182.529188   
Anhui Laian Baoshan Wind                      3670.613874  3670.521873   
Anhui Laian Longtougang Wind                  3682.866917  3682.775569   
Anhui Laian Longwosi Wind                     3672.393767  3672.301724   
Anning River Canyon (Phase 1) Dechang County  2680.444330  2680.607525   

SolarPowerPlant_Name                               Aksu C       Aksu D  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3182.306878  3182.371500   
Anhui Laian Baoshan Wind                      3670.378673  3670.484468   
Anhui Laian Longtougang Wind                  3682.632648  3682.738553   
Anhui Laian Longwosi Wind                     3672.158508  3672.264295   
Anning River Canyon (Phase 1) Dechang County  2680.581790  2680.720694   

SolarPowerPlant_Name                               Aksu E       Aksu F  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3183.209002  3184.188716   
Anhui Laian Baoshan Wind                      3671.248527  3672.521166   
Anhui Laian Longtougang Wind                  3683.502237  3684.775572   
Anhui Laian Longwosi Wind                     3673.028377  3674.300973   
Anning River Canyon (Phase 1) Dechang County  2681.277660  2682.705241   

SolarPowerPlant_Name                           Aksu Jinko     Ala'er A  ...  \
WindPowerPlant_Name                                                     ...   
Abag Banner Huiteng Liang Phase 1             3181.660867  3086.424381  ...   
Anhui Laian Baoshan Wind                      3669.320824  3559.211154  ...   
Anhui Laian Longtougang Wind                  3681.573685  3571.437685  ...   
Anhui Laian Longwosi Wind                     3671.100728  3560.992644  ...   
Anning River Canyon (Phase 1) Dechang County  2679.192771  2568.934766  ...   

SolarPowerPlant_Name                          Zhongwei 30 ZHW     Zhongxin  \
WindPowerPlant_Name                                                          
Abag Banner Huiteng Liang Phase 1                 1192.269778  2357.406846   
Anhui Laian Baoshan Wind                          1377.284041  1807.072647   
Anhui Laian Longtougang Wind                      1389.801896  1813.783124   
Anhui Laian Longwosi Wind                         1379.040257  1808.577669   
Anning River Canyon (Phase 1) Dechang County      1198.466757   177.164949   

SolarPowerPlant_Name                          Zhuanjingzhen A  \
WindPowerPlant_Name                                             
Abag Banner Huiteng Liang Phase 1                 1002.650319   
Anhui Laian Baoshan Wind                          1145.412855   
Anhui Laian Longtougang Wind                      1158.128095   
Anhui Laian Longwosi Wind                         1147.133301   
Anning River Canyon (Phase 1) Dechang County      1264.764095   

SolarPowerPlant_Name                          Zhuanjingzhen B  \
WindPowerPlant_Name                                             
Abag Banner Huiteng Liang Phase 1                 1000.747025   
Anhui Laian Baoshan Wind                          1143.034162   
Anhui Laian Longtougang Wind                      1155.751494   
Anhui Laian Longwosi Wind                 

**1.3. Seleccionar una fila de la matriz de distancias y graficas dos puntos con la mínima distancia sobre el mapa del país.**

Se ha seleccionado la fila correspondiente con la planta de energía eólica "Abag Banner Huiteng Liang Phase 1", se determinó a cuál planta de energía solar se encontraba más cerca y ambos puntos se graficaron en un mapa interactivo de China.

In [308]:
# Planta de energía solar más cercana a la planta eólica "Abag Banner Huiteng Liang Phase 1"
AbagBanner_ClosestSolarPlant = DistanceMatrix_WindSolarPlants.loc["Abag Banner Huiteng Liang Phase 1"].idxmin(axis=0)

#Ambos puntos:
Abag_Banner = China3415_WindPowerPlants[China3415_WindPowerPlants.WindPowerPlant_Name=="Abag Banner Huiteng Liang Phase 1"]
AbagBanner_ClosestSolarPlant_p = China3415_SolarPowerPlants[China3415_SolarPowerPlants.SolarPowerPlant_Name == AbagBanner_ClosestSolarPlant]

#Mapa interactivo con ambos puntos 
base = Abag_Banner.explore(color="blue",zoom_start =4, marker_kwds=dict(radius=3))
AbagBanner_ClosestSolarPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=3))

## Ejercicio 2
1. Use a map of points and a map of lines.

2. Compute the distance matrix for both maps.

3. Select one line of the distance matrix, and plot the closests and the farthest point to that line.

**2.1. Mapas de puntos y líneas a emplear:**
- Mapa de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants).

- Mapa de líneas: Arroyos en China (China3415_Streams).

Fuentes empleadas: [Global Power Plant Database](https://datasets.wri.org/dataset/globalpowerplantdatabase), [Stream Network dataset](https://data.apps.fao.org/catalog/iso/dc2a5121-0b32-482b-bd9b-64f7a414fa0d). 

In [293]:
China3415_NuclearPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Nuclear")

China3415_GeoMaps = "https://github.com/Ciencia-de-datos-espaciales-2023-2/Evaluacion4_1ra-Parte_Mining_GeoDFs/raw/main/maps/China_3415/China3415_GeoMap.gpkg"
China3415_Streams = gpd.read_file(China3415_GeoMaps,layer="Streams") 

**2.2. Matriz de distancias entre ambos mapas**

Nota: Ambos mapas presentan proyecciones en metros. Pero, la matriz de distancias se encuentra en unidades de kilómetros.

In [294]:
China3415_NuclearPowerPlants_n = China3415_NuclearPowerPlants.set_index("NuclearPowerPlant_Name")
China3415_Streams_n = China3415_Streams.set_index("Stream_UniqueArcIdentifier_ARCID")

DistanceMatrix_StreamNuclearPlant = China3415_Streams_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 5 primeras filas de la matriz de distancias:
DistanceMatrix_StreamNuclearPlant.head()

NuclearPowerPlant_Name                   CEFR  Changjiang    Daya Bay  \
Stream_UniqueArcIdentifier_ARCID                                        
147498                            2478.820456  148.071847  681.348000   
147474                            2475.493876  150.004222  675.789126   
147459                            2469.945882  161.353787  661.486652   
147485                            2476.286481  137.043392  681.026276   
147735                            2499.333836  151.246731  706.978115   

NuclearPowerPlant_Name            Fangchenggang  Fangjiashan       Fuqing  \
Stream_UniqueArcIdentifier_ARCID                                            
147498                               382.824449  1752.686233  1267.321384   
147474                               382.756267  1747.513190  1261.653503   
147459                               388.180682  1734.898603  1246.431720   
147485                               374.461390  1751.796322  1267.266474   
147735                               392.637375  1777.819473  1293.080905   

NuclearPowerPlant_Name              Hongyanhe     Ling Ao       Ningde  \
Stream_UniqueArcIdentifier_ARCID                                         
147498                            2667.572830  682.110064  1443.381637   
147474                            2663.397237  676.548675  1437.805301   
147459                            2654.756399  662.233545  1423.329896   
147485                            2665.663959  681.791246  1443.081926   
147735                            2690.488389  707.742054  1469.027512   

NuclearPowerPlant_Name              Qinshan 2      Tianwan   Yangjiang  
Stream_UniqueArcIdentifier_ARCID                                        
147498                            1752.767478  2059.427691  450.084480  
147474                            1747.592137  2055.032856  445.111214  
147459                            1734.967546  2045.575079  433.575587  
147485                            1751.880487  2057.710958  448.979210  
147735                            1777.904532  2082.909816  474.912826

**2.3. Seleccionar una fila de la matriz de distancias y graficar el punto/línea más cercano y lejano.**

Se ha seleccionado la fila correspondiente al arroyo con identificador ARCID 146955, se determinó a cuál planta de energía nuclear se encontraba más cerca y lejos. Los dos puntos y la línea se graficaron en un mapa interactivo de China.

Nota: El arroyo se ve como una línea muy pequeña, porque su masa de agua es menor comparada con la de un río.

In [309]:
# Planta de energía nuclear más cercana y lejana al arroyo con identificador ARCID 146955
Stream146955_ClosestNuclearPlant = DistanceMatrix_StreamNuclearPlant.loc[146955].idxmin(axis=0)
Stream146955_FurthestNuclearPlant = DistanceMatrix_StreamNuclearPlant.loc[146955].idxmax(axis=0)

#Coordenadas de ambos puntos y el arroyo con identificador ARCID 146955
Stream146955_ClosestNuclearPlant_p = China3415_NuclearPowerPlants[China3415_NuclearPowerPlants.NuclearPowerPlant_Name==Stream146955_ClosestNuclearPlant]
Stream146955_FurthestNuclearPlant_p = China3415_NuclearPowerPlants[China3415_NuclearPowerPlants.NuclearPowerPlant_Name==Stream146955_FurthestNuclearPlant]
Stream146955_p = China3415_Streams[China3415_Streams.Stream_UniqueArcIdentifier_ARCID==146955]

#Mapa interactivo con ambos puntos y el arroyo con identificador ARCID 146955
base = Stream146955_p.explore(color="blue",zoom_start=4)
Stream146955_ClosestNuclearPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=3))
Stream146955_FurthestNuclearPlant_p.explore(m=base, color="green",marker_kwds=dict(radius=3))

## Ejercicio 3   
1. Create a set of points and a set of polygons

2. Compute the distance matrix for both sets.

3. Select one polygon of the distance matrix, and plot the closests and the farthest point to that polygon.

**3.1. Set de puntos y polígonos a emplear:**
- Set de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants).Es el mismo que se empleó en la pregunta 2.

- Set de polígonos: Convex Hulls de arroyos en China que pertenecen a una misma cuenca principal (China3415_MajorStreamBasin_Hulls). El GeoDataFrame con los convex hulls ya fue creado en el Jupyter Notebook de preprocesamiento.


In [296]:
China3415_MajorStream_BasinsHulls = gpd.read_file(China3415_GeoMaps,layer="MajorStream_BasinsHulls") 

**3.2. Matriz de distancias entre ambos mapas**

Nota: La matriz de distancias se encuentra en unidades de kilómetros.

In [297]:
China3415_MajorStream_BasinsHulls_n = China3415_MajorStream_BasinsHulls.set_index("Major_Basin")

DistanceMatrix_StreamBasinsHulls_NuclearPlant = China3415_MajorStream_BasinsHulls_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 5 primeras filas de la matriz de distancias:
DistanceMatrix_StreamBasinsHulls_NuclearPlant.head()

NuclearPowerPlant_Name                   CEFR   Changjiang     Daya Bay  \
Major_Basin                                                               
Amur                               679.388929  3009.594588  2438.981415   
Bo Hai - Korean Bay, North Coast   228.475546  2472.452237  1966.415288   
China Coast                        392.981428   620.417018     0.000000   
Ganges - Bramaputra               2120.404089  1532.320412  1832.754215   
Gobi Interior                      269.736509  2070.761509  1936.420574   

NuclearPowerPlant_Name            Fangchenggang  Fangjiashan       Fuqing  \
Major_Basin                                                                 
Amur                                2811.009362  1397.413050  1973.975974   
Bo Hai - Korean Bay, North Coast    2262.261451   997.730963  1569.474959   
China Coast                          533.166772     0.000000    16.651114   
Ganges - Bramaputra                 1351.292290  2284.688847  2201.308181   
Gobi Interior                       1825.379265  1429.137229  1872.212425   

NuclearPowerPlant_Name              Hongyanhe      Ling Ao       Ningde  \
Major_Basin                                                               
Amur                               413.208049  2438.573933  1780.045817   
Bo Hai - Korean Bay, North Coast     0.000000  1966.199124  1381.830381   
China Coast                        238.812795     0.000000     0.000000   
Ganges - Bramaputra               2564.211323  1833.684551  2248.927315   
Gobi Interior                      564.656228  1936.863642  1746.612206   

NuclearPowerPlant_Name              Qinshan 2      Tianwan    Yangjiang  
Major_Basin                                                              
Amur                              1397.885615  1012.444800  2627.316040  
Bo Hai - Korean Bay, North Coast   998.415499   547.874727  2120.878176  
China Coast                          0.000000     0.000000   191.334274  
Ganges - Bramaputra               2285.665828  2202.074934  1670.965223  
Gobi Interior                     1430.261066   933.420035  1938.894073

**3.3. Seleccionar un polígono y graficar el punto más cercano y lejano.**

Se ha seleccionado la fila correspondiente al polígono de la cuenca "Huang He", se determinó a cuál planta de energía nuclear se encontraba más cerca y lejos. Los dos puntos y el polígono se graficaron en un mapa interactivo de China.

In [310]:
# Planta de energía nuclear más cercana y lejana al polígono de la cuenca "Huang He"
HuangHeHull_ClosestNuclearPlant = DistanceMatrix_StreamBasinsHulls_NuclearPlant.loc["Huang He"].idxmin(axis=0)
HuangHeHull_FurthestNuclearPlant = DistanceMatrix_StreamBasinsHulls_NuclearPlant.loc["Huang He"].idxmax(axis=0)

#Coordenadas de ambos puntos y el polígono
HuangHeHull_ClosestNuclearPlant_p = China3415_NuclearPowerPlants[China3415_NuclearPowerPlants.NuclearPowerPlant_Name==HuangHeHull_ClosestNuclearPlant]
HuangHeHull_FurthestNuclearPlant_p = China3415_NuclearPowerPlants[China3415_NuclearPowerPlants.NuclearPowerPlant_Name==HuangHeHull_FurthestNuclearPlant]
HuangHeHull = China3415_MajorStream_BasinsHulls[China3415_MajorStream_BasinsHulls.Major_Basin=="Huang He"]

#Mapa interactivo con ambos puntos y el polígono
base = HuangHeHull.explore(color="blue",zoom_start=3)
HuangHeHull_ClosestNuclearPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=4))
HuangHeHull_FurthestNuclearPlant_p.explore(m=base, color="green",marker_kwds=dict(radius=4))

## Ejercicio 4 
1. Create a set of points and a set of lines

2. Get the buffer for the lines, select different values for the distance.

3. Keep the points that are within the buffer (as in point 2, you need to play with differn distances until you show something interesting).  

**4.1. Set de puntos y polígonos a emplear:**
- Set de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants). Es el mismo que se empleó en las preguntas 2 y 3.

- Set de líneas: Conjunto de arroyos en China que pertenecen a una misma cuenca principal (China3415_MajorStreamBasin). El GeoDataFrame con las multilíneas correspondientes a un cuenca principal ya fue creado en el Jupyter Notebook de preprocesamiento.


In [299]:
China3415_MajorStream_Basins = gpd.read_file(China3415_GeoMaps,layer="MajorStream_Basins") 

**4.2. *Buffers* alrededor de la cuenca China Coast**

Se crearon *buffers* de los arroyos que forman parte de la cuenca "China Coast". Se seleccionó esta cuenca, porque se localiza al sureste de China, donde la mayoría de las plantas nucleares también se encuentran.

Primero, con la finalidad de evaluar qué rango de distancias emplear para crear los *buffers*, se creó una matriz de distancias:

In [300]:
China3415_MajorStream_Basins_n = China3415_MajorStream_Basins.set_index("Major_Basin")
DistanceMatrix_StreamBasins_NuclearPlant = China3415_MajorStream_Basins_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

import pandas as pd
ChinaCoast_Distances = pd.Series(list(DistanceMatrix_StreamBasins_NuclearPlant.loc["China Coast"]))
ChinaCoast_Distances.describe()

count     12.000000
mean     173.754462
std      227.193252
min        4.340807
25%       12.163869
50%       20.341937
75%      280.885002
max      620.417018
dtype: float64

Con los resultados obtenidos de esta evaluación preliminar, se observa que la planta nuclear más cercana a la cuenca "Coast China" está a 4.34 km, mientras que la más lejana, a 620.42 km. Con esto en consideración, se crearon buffers de 5, 10 y 25 km:

Nota: No fue posible trabajar con mayores distancias, porque el tiempo de computación requerido iba a ser demasiado elevado.

In [301]:
ChinaCoastBuffer_5km = China3415_MajorStream_Basins[China3415_MajorStream_Basins.Major_Basin=="China Coast"
                                                   ].buffer(distance = 5*1000)

ChinaCoastBuffer_10km = China3415_MajorStream_Basins[China3415_MajorStream_Basins.Major_Basin=="China Coast"
                                                    ].buffer(distance = 10*1000)

ChinaCoastBuffer_25km = China3415_MajorStream_Basins[China3415_MajorStream_Basins.Major_Basin=="China Coast"
                                                    ].buffer(distance = 25*1000)

**4.3. Mantener los puntos que se encuentran dentro de cada *buffer* y verificar con cuál se obtienen los resultados más interesantes**

Se determinó el número de plantas de energía nuclear dentro de cada buffer. Con esto, se concluyó que, fijando un demarcación de 25 km en torno a los arroyos que se encuentran dentro de la cuenca "Coast China", 7 de las 12 plantas nucleares en China se encuentran dentro del buffer. A 5 km y 10 km, se obtiene un menor número:

In [302]:
NuclearPowerPlants_WithinBuffer_5km = China3415_NuclearPowerPlants.clip(mask=ChinaCoastBuffer_5km)
NuclearPowerPlants_WithinBuffer_10km = China3415_NuclearPowerPlants.clip(mask=ChinaCoastBuffer_10km)
NuclearPowerPlants_WithinBuffer_25km = China3415_NuclearPowerPlants.clip(mask=ChinaCoastBuffer_25km)

Resultados = {
    "5km" : len(set(NuclearPowerPlants_WithinBuffer_5km.NuclearPowerPlant_Name)),
    "10km" : len(set(NuclearPowerPlants_WithinBuffer_10km.NuclearPowerPlant_Name)),
    "25km" : len(set(NuclearPowerPlants_WithinBuffer_25km.NuclearPowerPlant_Name))}

Resultados

{'5km': 1, '10km': 3, '25km': 7}

Con estos resultados, se escogió el *buffer* a 25 km de la cuenca de "China Coast" para graficar los puntos que se encuentran dentro de dicho polígono:

Nota: En el mapa, solo se ven, aparentemente 5 puntos, porque las plantas nucleares Daya Bay y Ling Ao se encuentran muy próximas entre sí. Lo mismo ocurre con las plantas Qinshan 2 y Fangjiashan.

In [303]:
ChinaCoast = China3415_MajorStream_Basins[China3415_MajorStream_Basins.Major_Basin=="China Coast"]

base = ChinaCoast.explore(color="blue",zoom_start=5)
ChinaCoastBuffer_25km.explore(m=base, color="orange")
NuclearPowerPlants_WithinBuffer_25km.explore(m=base, color="red",marker_kwds=dict(radius=5))